plugins/reindex/src/main/java/org/elasticsearch/plugin/reindex/TransportReindexAction.java

```diff
@@ -0,0 +1,230 @@
+            if (sourceIndex.equals(target)) {
+                ActionRequestValidationException e = new ActionRequestValidationException();
+                e.addValidationError("index-by-search cannot write into an index its reading from [" + target + ']');
+                throw e;
+            }
+        }
+        new AsyncIndexBySearchAction(request, listener).start();
+    }
+
+    /**
+     * Simple implementation of reindex using scrolling and bulk. There are tons
+     * of optimizations that can be done on certain types of reindex requests
+     * but this makes no attempt to do any of them so it can be as simple
+     * possible.
+     */
+    class AsyncIndexBySearchAction extends AbstractAsyncBulkIndexByScrollAction<ReindexRequest, ReindexResponse> {
+        public AsyncIndexBySearchAction(ReindexRequest request, ActionListener<ReindexResponse> listener) {
+            super(logger, scriptService, client, threadPool, request, request.source(), listener);
+        }
+
+        @Override
+        protected IndexRequest buildIndexRequest(SearchHit doc) {
+            IndexRequest index = new IndexRequest(mainRequest.destination(), mainRequest);
+
+            // We want the index from the copied request, not the doc.
+            index.id(doc.id());
+            if (index.type() == null) {
+                /*
+                 * Default to doc's type if not specified in request so its easy
+                 * to do a scripted update.
+                 */
+                index.type(doc.type());
+            }
+            index.source(doc.sourceRef());
+            /*
+             * We've already copied the versionType but we need to be careful
+             * with internal versioning. It either represents "create" (version
+             * == MATCH_DELETED) or overwrite (version == MATCH_ANY). All other
+             * values are invalid.
+             */
+            if (index.versionType() == INTERNAL) {
+                if (index.version() != Versions.MATCH_DELETED && index.version() != Versions.MATCH_ANY) {
+                    throw new IllegalArgumentException("Invalid version for reindex with internal versioning [" + index.version() + ']');
+                }
+            } else {
+                index.version(doc.version());
+            }
+            return index;
+        }
+
+        /**
+         * Override the simple copy behavior to allow more fine grained control.
+         */
+        @Override
+        protected void copyRouting(IndexRequest index, SearchHit doc) {
+            String routingSpec = mainRequest.destination().routing();
+            if (routingSpec == null) {
+                super.copyRouting(index, doc);
+                return;
+            }
+            if (routingSpec.startsWith("=")) {
+                index.routing(mainRequest.destination().routing().substring(1));
+                return;
+            }
+            switch (routingSpec) {
+            case "keep":
```

Comment:
Is this formatting intended (`case` on same indentation level as enclosing `switch`)?